sdk/src/dlob/DLOB.ts

```diff
@@ -339,43 +348,102 @@ export class DLOB {
 				break;
 			}
 
-			const takerIsMaker =
-				crossingNodes?.makerNode !== undefined &&
-				crossingNodes.node.userAccount.equals(
-					crossingNodes.makerNode.userAccount
-				);
+			for (const crossingNode of crossingNodes) {
+				nodesToFill.push(crossingNode);
+			}
+		}
+		return nodesToFill;
+	}
+
+	public findvAMMCrossingNodesToFill(
+		marketIndex: number,
+		vBid: BN,
+		vAsk: BN,
+		slot: number,
+		marketType: MarketType,
+		oraclePriceData: OraclePriceData
+	): NodeToFill[] {
+		const askVAMM = getVammNodeGenerator(vAsk).next();
+		if (askVAMM.done) {
+			throw new Error('no vAMM ask node');
+		}
+		const bidVAMM = getVammNodeGenerator(vBid).next();
+		if (bidVAMM.done) {
+			throw new Error('no vAMM bid node');
+		}
+
+		const nodesToFill = new Array<NodeToFill>();
+
+		const askGenerator = this.getAsks(
+			marketIndex,
+			undefined, // dont include vask
+			slot,
+			marketType,
+			oraclePriceData
+		);
+		const bidGenerator = this.getBids(
+			marketIndex,
+			undefined, // dont include vbid
+			slot,
+			marketType,
+			oraclePriceData
+		);
+
+		let nextAsk = askGenerator.next();
+		let nextBid = bidGenerator.next();
+
+		// check for asks that cross vBid
+		while (!nextAsk.done) {
+			const askNode = nextAsk.value;
+			const askPrice = askNode.getPrice(oraclePriceData, slot);
 
-			// Verify that each side is different user
-			if (crossingNodes && !takerIsMaker) {
-				nodesToFill.push(crossingNodes);
+			if (askPrice.lte(vBid) && isAuctionComplete(askNode.order, slot)) {
+				nodesToFill.push({
+					node: askNode,
+					makerNode: bidVAMM.value,
+				});
+			} else {
```

Comment:
assuming the asks/bids are sorted by price, you can just break in this else block to avoid needing to traverse all the nodes?